examples/face_recognition_11/face_recognition_11/evaluator.py

```diff
@@ -0,0 +1,400 @@
+
+    fnmr_y = list()
+    fmr_y = list()
+
+    for threshold in thresholds:
+        tsm_for_one_threshold = [
+            compute_per_sample(gt, inf, threshold, configuration) for gt, inf in zip(ground_truths, inferences)
+        ]
+
+        tcm = compute_test_case_metrics(test_samples, ground_truths, tsm_for_one_threshold, 0, unique_pairs)
+        fnmr_y.append(tcm.FNMR)
+        fmr_y.append(tcm.FMR)
+
+    plots.append(
+        CurvePlot(
+            title="Recognition: Test Case FNMR vs. Baseline FMR",
+            x_label="Baseline False Match Rate",
+            y_label="Test Case False Non-Match Rate (%)",
+            x_config=AxisConfig(type="log"),
+            curves=[Curve(x=baseline_fmr_x, y=fnmr_y, extra=dict(Threshold=thresholds))],
+        )
+    )
+
+    plots.append(
+        CurvePlot(
+            title="Recognition: Test Case FMR vs. Baseline FMR",
+            x_label="Baseline False Match Rate",
+            y_label="Test Case False Match Rate",
+            x_config=AxisConfig(type="log"),
+            y_config=AxisConfig(type="log"),
+            curves=[Curve(x=baseline_fmr_x, y=fmr_y, extra=dict(Threshold=thresholds))],
+        )
+    )
+
+    plots.append(create_similarity_histogram(test_samples, ground_truths, inferences))
+
+    return plots
+
+
+def compute_test_suite_metrics(
+    metrics: List[TestSampleMetrics], baseline: TestCaseMetrics, threshold: float
+) -> TestSuiteMetrics:
+    return TestSuiteMetrics(
+        Threshold=threshold,
+        FM=baseline.FM,
+        FNM=baseline.FNM,
+        FNMR=baseline.FNM / baseline.nGenuinePairs,
+        TotalFTE=np.sum([tsm.bbox_failure_to_enroll or tsm.keypoint_failure_to_align for tsm in metrics]),
+        TotalBBoxFTE=np.sum([tsm.bbox_failure_to_enroll for tsm in metrics]),
+        TotalKeypointFTE=np.sum([tsm.keypoint_failure_to_align for tsm in metrics]),
+    )
+
+
+def evaluate_face_recognition_11(
+    test_samples: List[TestSample],
+    ground_truths: List[GroundTruth],
+    inferences: List[Inference],
+    test_cases: TestCases,
+    configuration: ThresholdConfiguration,
+) -> EvaluationResults:
+    pairs = set([(ts.locator, pair.locator) for ts in test_samples for pair in ts.pairs])
+    unique_pairs = set((a, b) if a <= b else (b, a) for a, b in pairs)
+
+    threshold = compute_threshold(test_samples, ground_truths, inferences, configuration.false_match_rate)
+
+    # compute per-sample metrics for each test sample
+    test_sample_metrics = [
+        compute_per_sample(gt, inf, threshold, configuration, test_samples[i])
+        for i, (gt, inf) in enumerate(zip(ground_truths, inferences))
+    ]
+
+    # compute aggregate metrics across all test cases using `test_cases.iter(...)`
+    all_test_case_metrics: List[Tuple[TestCase, TestCaseMetrics]] = []
+    all_test_case_plots: List[Tuple[TestCase, List[Plot]]] = []
+    baseline = compute_test_case_metrics(test_samples, ground_truths, test_sample_metrics, 0, unique_pairs)
```

Comment:
It looks like there are many places where `compute_test_case_metrics` is called just to calculate FMR and FNMR.  This will results in multiple computation of not only the FR metrics but also the detection metrics.  You can make the FR metrics computation a separate function and it can be called within `compute_test_case_metrics` and in this case, you don't perform unnecessary computation.

Reply:
We can also move `unique_pairs` computation inside this new `compute_pair_metrics` function.